### PR TITLE
make paldb reader thread safe

### PR DESCRIPTION
**Pull request by xiaoxxcool**
I made some efforts to make paldb reader works in multi-thread environment, the perfermance decreased about 20% in single thread read, the perfermance test report is

```
Before change:
FILE LENGTH;	KEYS;		RPS
1831		100		3291755
16828		1000		2999658
166741		10000		2778105
1664749		100000		2019042
16643345	1000000		1587133
166431370	10000000	1424029

After change:
FILE LENGTH;	KEYS;		RPS
1831		100		2619438
16828		1000		2394381
166741		10000		2167783
1664749		100000		1577364
16643345	1000000		1388372
166431370	10000000	1192194


```
the benefit is you can share the StoreReader instance without lock.

### CODE REVIEW

**Comment by bowenli86** [comment]
Just curious. Isn't StoreReader read only? Why does it need to be thread safe?

**Comment by xiaoxxcool** [comment]
my server contains some data update periodically, each request is processed in different threads, when the data is taken from the shared StoreReader instance, lock is required, or it will get the wrong data. what is a good practice to use paldb in such circumstance?

File: paldb/src/main/java/com/linkedin/paldb/impl/StorageReader.java [hunk 1]
```diff
@@ -83,9 +85,6 @@
   private FileChannel channel;
   // Use MMap for data?
   private final boolean mMapData;
-  // Buffers
-  private final DataInputOutput sizeBuffer = new DataInputOutput(new byte[5]);
-  private final byte[] slotBuffer;
```
Comment (scottcarey) [inline paldb/src/main/java/com/linkedin/paldb/impl/StorageReader.java:88]: Using a ThreadLocal<byte[]> (perhaps static, with max size being used for all slot sizes) would improve performance, reducing allocation.   The same goes for the `DataInputOutput`s.   Perhaps a single class, containing all of the necessary buffers that a thread needs (arrays, DataInputOutputs, Hash instance, etc) and a single ThreadLocal holding the per thread buffers would perform best.

File: paldb/src/main/java/com/linkedin/paldb/impl/StorageReader.java [hunk 2]
```diff
@@ -350,7 +350,7 @@ public int getKeyCount() {
   }
 
   //Get data from disk
-  private byte[] getDiskBytes(long offset)
+  private synchronized byte[] getDiskBytes(long offset)
```
Comment (scottcarey) [inline paldb/src/main/java/com/linkedin/paldb/impl/StorageReader.java:353]: Does this need to be synchronized?

RandomAccessFile is slow anyway.  It is synchronized.   SeekableFileChannel is faster, and a memory mapped file is faster than that.  Perhaps changing to one of those would be better.

**Pull request closed without merging**